db/range_del_aggregator.h

```diff
@@ -166,6 +164,7 @@ class RangeDelAggregator {
   struct Rep {
     StripeMap stripe_map_;
     PinnedIteratorsManager pinned_iters_mgr_;
+    std::list<std::string> pinned_slices_;
```

Comment:
After debugging I believe the corruption can only happen if `UpdateInternalKey` is called or if `block_restart_intervals != 1`, as either of those prevent the map from holding pointers to the full keys inside the range deletion meta-block (which the `PinnedIteratorsManager` should be capable of pinning). `UpdateInternalKey` is only called for range deletion keys as of this PR, and we always set `block_restart_intervals = 1`. I wonder if there's a way to prevent the file ingestion path from using `UpdateInternalKey`, rather than maintaining this list for all range deletion keys (even though it's only needed for ones added through the file ingestion path).

Reply:
Thanks for looking into this and thanks for adding an assertion to guard against this in 7b65521.

As you brought up, the file ingestion path uses `UpdateInternalKey` because it creates blocks with global sequence numbers. This is used to avoid needing to update every single key in an SST when a sequence number is computed during ingestion time. I see two general approaches to addressing this issue, but I'm not confident about which one would be the better approach:

The first is that we look into addressing the following TODO:
https://github.com/facebook/rocksdb/blob/aa53579d6cc5a53c1493bbda4f21c0e9831b377b/table/block.cc#L256-L257
If we updated the sequence number in the loaded block directly in `BlockIter::ParseNextKey` when we see a global sequence number then we wouldn't ever need to un-pin the keys. This would allow us to avoid needing to worry about making any changes to `RangeDelAggregator`. However, there are a few downsides here, including that this could alter performance characteristics outside of this change. I also don't feel comfotable making a change like that without some guidance because I expect that it could have far reaching consequences.

The other approach, mentioned by @anand1976 below, is that we allow unpinned keys in `RangeDelAggregator::AddTombstones` but we only perform the copy if a key is unpinned. This has a few nice properties, including that it is a well contained change, it shouldn't have any unexpected consequences or performance implications, and it adds a bit more flexibility to `RangeDelAggregator` in case another future user of the class provides an iterator with unpinned keys. This is the direction I'm leaning towards. Please let me know if you agree. Also, if so, could you give me a pointer about how this kind of memory pinning usually takes place in this code base? The std::list was just for prototyping, but it certainly isn't the best way to do this kind of thing. Is there some kind of arena object that I should be copying the keys into?

Reply:
In approach #1, it looks like there is no guarantee that the sequence number can be updated in the loaded block directly, since the key may be sharing bytes with the previous key and would have to be copied anyway. So I suggest going with approach#2. @ajkr Do you agree?

As far as pinning is concerned, maybe rep_->pinned_slices_.emplace_back(input->key().data(), input->key().size()) would be more efficient. Also, pinned_slices_ could be std::vector or std::deque instead of std::list to avoid repeated allocations.

Reply:
Sorry for the delay. Yeah, approach 2 sounds safer to me too. And if we decide to do approach 1 later, you'll get the benefit of avoiding the copy without having to change anything.

I don't have a concrete idea of a problem for approach 1. Regarding shared bytes, we don't delta encode keys in range deletion meta-blocks so it shouldn't be a problem.

Reply:
Good catch on the `IsValuePinned()` recent change